yt/frontends/halo_catalog/data_structures.py

```diff
@@ -44,14 +44,68 @@ def _setup_filenames(self):
               [cls(self.dataset, self.io,
                    self.dataset.parameter_filename, 0)]
 
-class HaloCatalogHDF5File(ParticleFile):
+class HaloCatalogFile(ParticleFile):
+    def __init__(self, ds, io, filename, file_id):
+        self._coords = {}
+        super(HaloCatalogFile, self).__init__(
+            ds, io, filename, file_id)
+
+    def _read_particle_positions(self, ptype, f=None):
+        raise NotImplementedError
+
+    def _get_particle_positions(self, ptype, f=None):
+        if ptype in self._coords:
+            return self._coords[ptype]
+
+        pcount = self.total_particles[ptype]
+        cache_pos = getattr(self.ds, "cache_positions", False)
+        if pcount == 0:
+            if cache_pos:
+                self._coords[ptype] = None
+            return None
+
+        # These are 32 bit numbers, so we give a little lee-way.
+        # Otherwise, for big sets of particles, we often will bump into the
+        # domain edges.  This helps alleviate that.
+        dle = self.ds.domain_left_edge.to('code_length').v
+        dre = self.ds.domain_right_edge.to('code_length').v
+        dx = 2. * np.finfo(np.float32).eps
+        pos = self._read_particle_positions(ptype, f=f)
+        np.clip(pos, dle + dx, dre - dx, pos)
```

Comment:
I admit I'm not totally following this, but are we respecting the "only read in these bounds" rules that we allow in the constructor to the dataset object?

Reply:
I don't understand this comment. What are these rules you're referring to?

This clipping strategy was code that appeared in `_initialize_index` for each of the frontends here. It seemed like a good idea to apply the same logic to the actual positions, but maybe it doesn't matter.

Reply:
I think he's referring to supplying a `bounding_box` keyword argument to `yt.load()`:

```
In [1]: ds.domain_left_edge
Out[1]: YTArray([ -547.51629822,  -557.30091064, -3675.19602173]) code_length

In [2]: ds.domain_right_edge
Out[2]: YTArray([ 579.153078  ,  472.81018799, 1591.69712036]) code_length

In [3]: ad = ds.all_data()

In [4]: ad['gas', 'density'].shape
Out[4]: (169240,)

In [6]: ds = yt.load('TipsyGalaxy/galaxy.00300', bounding_box=[ds.domain_left_edge/2, ds.domain_right_edge/2])

In [7]: ds.domain_left_edge
Out[7]: YTArray([ -273.75814911,  -278.65045532, -1837.59801086]) code_length

In [8]: ds.domain_right_edge
Out[8]: YTArray([289.576539  , 236.40509399, 795.84856018]) code_length

In [9]: ad = ds.all_data()

In [10]: ad['gas', 'density'].shape
Out[10]: (167728,)
```